src/main/scala/com/hungry/cars/domain/error/UserError.scala

```diff
@@ -0,0 +1,72 @@
+package com.hungry.cars.domain.error
+
+import io.circe.Encoder
+import io.circe.Json
+import io.circe.syntax.EncoderOps
+
+sealed trait UserError {
+  def code: String
+  def errorMessage: String
+}
+
+object UserError {
+
+  case class UserAlreadyExists(userName: String) extends UserError {
+    def code: String         = "user-error-001"
+    def errorMessage: String = s"Username $userName already exists!"
+  }
+
+  case class UsernameHasSpecialCharacters(userName: String) extends UserError {
+    def code: String         = "user-error-002"
+    def errorMessage: String = s"Username $userName cannot contain special characters."
+  }
+
+  case class EmailIsNotValid(email: String) extends UserError {
+    def code: String         = "user-error-003"
+    def errorMessage: String = "Email is not valid!"
+  }
+
+  case class EmailAlreadyExists(email: String) extends UserError {
+    def code: String         = "user-error-004"
+    def errorMessage: String = s"Email $email already exists!"
+  }
+
+  case object PasswordDoesNotMeetCriteria extends UserError {
+    def code: String = "user-error-005"
+
+    def errorMessage: String =
+      "Password must be at least 8 characters long, including an uppercase, a lowercase letter and one number."
+  }
+
+  case object PasswordsAreNotIdentical extends UserError {
+    def code: String         = "user-error-006"
+    def errorMessage: String = "Passwords are not identical!"
+  }
+
+  case object PasswordCanNotEncrypt extends UserError {
+    def code: String         = "user-error-007"
+    def errorMessage: String = "Password can not encrypt!"
+  }
+
+  case class FirstNameHasSpecialCharacters(firstName: String) extends UserError {
+    def code: String         = "user-error-008"
+    def errorMessage: String = s"First name $firstName cannot contain spaces, numbers or special characters."
+  }
+
+  case class LastNameHasSpecialCharacters(lastName: String) extends UserError {
+    def code: String         = "user-error-009"
+    def errorMessage: String = s"Last name $lastName cannot contain spaces, numbers or special characters."
+  }
+
+  case object AgeIsInvalid extends UserError {
+    def code: String         = "user-error-010"
+    def errorMessage: String = "You must be aged 18 and not older than 75 to use our services."
+  }
+
+  private def toJson(userError: UserError): Json = {
```

Comment:
To jest okej, ale w razie czego istnieje taka biblioteka jak circe-tagged (https://index.scala-lang.org/abdolence/circe-tagged-adt-codec/circe-tagged-adt-codec/0.10.0?target=_3.x). Na razie to zostaw w ten sposób, ale warto wiedzieć o takiej bibliotece ;)